### ISSUE TITLE
Make it possible to force linewise operation on custom text objects

### ISSUE THREAD

**Issue by wellle** [vim/vim#3490]
For (most) built in text objects it's possible to force operation on them to be linewise, for example by using `dVab` (`:h o_V`, `motion_force`). When using custom text objects (defined as mappings by plugins for example), this doesn't currently work.

For example, put this in your `.vimrc`:
```vim
onoremap x viw
```
Open any file with a few lines each containing some words. With the cursor on any word, try these:
1. `dw` (builtin) deletes some characters
2. `dVw` (builtin) deletes linewise
3. `dx` (from mapping) deletes some characters
4. `dVx` (from mapping) deletes some characters, but should delete linewise

Vim should fix that and operate linewise when the operation was invoked with `o_V`.

Some notes: As far as I'm aware plugins use visual mode to select text objects in operator-pending mode. This is done either in visual mode `:normal v` or visual line mode `:normal V`. I'd expect neovim to be able to recognize whether these selections were done linewise or not. Then the user can force the motion `dvab`, `dVab` or not `dab`. If not forced, the mode used by the plugin should be used. If forced, the selection might need to be changed accordingly before applying the pending operator.

For reference, this came up in https://github.com/wellle/targets.vim/issues/214 and I already opened the same issue for neovim in https://github.com/neovim/neovim/issues/8667.

Thank you!

**Comment by andymass** [comment]
With `onoremap x viw`, `dVx` is like saying `dVviw`, which is why it doesn't work.

If you use `onoremap x iw`, then forcing with `dVx` actually _does_ work.

More precisely, a `v` in this place does not refer to the "visual mode" style of omaps.  If you can make your map in a way that doesn't require `:normal v` then it basically works.  This means positioning the cursor and not entering visual mode.  For example, this seems correct with forcing;

    onoremap x :call cursor(line('.')+1,col('.')+1)<cr>

But many custom maps can't be written this way and it's very difficult to make such maps behave properly with forcing, such as:

    onoremap x :normal viw<cr>

Note this has also been discussed at the following:
https://groups.google.com/forum/#!msg/vim_dev/MNtX3jHkNWw/bhL-o6TlKtgJ
https://groups.google.com/forum/#!msg/vim_dev/lR5rONDwgs8/iLsVCrxo_WsJ
https://github.com/kana/vim-textobj-user/issues/17

I personally would like such a `v:motiontype` or similar variable to be added as it would simplify considerably maps that aim to behave correctly in each forcing case.  Currently I resort to using four omaps e.g., `x`, `vx`, `Vx`, `<c-v>x`.

**Comment by wellle** [comment]
@andymass: Thank you! Yes the example `onoremap x viw` was just given to keep in simple here. As you said, in practice we are concerned about mappings we can currently only be implemented using visual mode. And in that case the plugin currently can't see whether the user wanted to force a motion type, nor does Vim handle that case on its own.

Thank you for the links above. I agree that `v:motiontype` or similar would be enough to make it possible to handle this on the plugin side. And we already do have a patch proposed. If we're still blocked on a name, I'd be fine with `v:force_motiontype`.

It's a bit sad to see the patch sitting there for 10 years now just because we can't pick a name. @glts even offered to update it accordingly 5 years ago. Is there anything we can do to make this move forward?

Thanks again 👍

**Comment by brammool** [comment]
> With `onoremap x viw`, `dVx` is like saying `dVviw`, which is why it doesn't work.
> 
> If you use `onoremap x iw`, then forcing with `dVx` actually _does_ work.
> 
> More precisely, a `v` in this place does not refer to the "visual mode" style of omaps.  If you can make your map in a way that doesn't require `:normal v` then it basically works.  This means positioning the cursor and not entering visual mode.  For example, this seems correct with forcing;
> 
>     onoremap x :call cursor(line('.')+1,col('.')+1)<cr>
> 
> But many custom maps can't be written this way and it's very difficult
> to make such maps behave properly with forcing, such as:
> 
>     onoremap x :normal viw<cr>

That's not a good example, since as you mentioned, leaving out the "v"
makes it work.  So what is an example where you do need the "v"?

> Note this has also been discussed at the following:
> https://groups.google.com/forum/#!msg/vim_dev/MNtX3jHkNWw/bhL-o6TlKtgJ
> https://groups.google.com/forum/#!msg/vim_dev/lR5rONDwgs8/iLsVCrxo_WsJ
> https://github.com/kana/vim-textobj-user/issues/17
> 
> I personally would like such a `v:motiontype` or similar variable to
> be added as it would simplify considerably maps that aim to behave
> correctly in each forcing case.  Currently I resort to using four
> omaps e.g., `x`, `vx`, `Vx`, `<c-v>x`.

-- 
hundred-and-one symptoms of being an internet addict:
115. You are late picking up your kid from school and try to explain
     to the teacher you were stuck in Web traffic.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by andymass** [comment]
@brammool  

The example I meant to draw attention to is `onoremap x :normal viw<cr>`, which is how most complex omaps are written, and all custom text objects are implemented.  In this case `dVx`, `dvx`, and `d<c-v>x` do not work properly.

**Comment by wellle** [comment]
@brammool: Here's a more realistic example which defines `i-` to operate between `-` characters. Put this in your .vimrc:
```vim
onoremap i- :normal T-vt-<CR>
```
Now open a file with contents like this
```
aaa - bbb x ccc - ddd
```
With the cursor on `x`, if you type `di-` it should leave you with
```
aaa -- ddd
```
So far so good. But if you type `dVi-` it should actually delete the full line (linewise), but doesn't. It behaves as if you typed `di-`.

**Comment by brammool** [comment]
> @brammool: Here's a more realistic example which defines `i-` to operate between `-` characters. Put this in your .vimrc:
> ```vim
> onoremap i- :normal T-vt-<CR>
> ```
> Now open a file with contents like this
> ```
> aaa - bbb x ccc - ddd
> ```
> With the cursor on `x`, if you type `di-` it should leave you with
> ```
> aaa -- ddd
> ```
> So far so good. But if you type `dVi-` it should actually delete the full line (linewise), but doesn't. It behaves as if you typed `di-`.

I can see where you are going, but you would need to search in multiple
lines to actually make it useful, otherwise one would just type "dd".

	onoremap i- :normal ?-<CR>v/-<CR>

I don't think you need anything new.  You can use mode() to check if you
are already in Visual mode.  If so, move the start of the visual area
and move the cursor to the end.  Otherwise just use "v" to start Visual
mode.  Something like:

	onoremap <expr> i- GetCommand()
	func GetCommand()
	  if mode() =~ '^[vV\<C-V>]'
	     normal ?-<CR>o/-<CR>
	  else
	     normal ?-<CR>v/-<CR>
	  endif
	  return ''
	endfunc

Haven't actually tried it.

-- 
hundred-and-one symptoms of being an internet addict:
116. You are living with your boyfriend who networks your respective
     computers so you can sit in separate rooms and email each other

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by wellle** [comment]
@brammool: I can't get your example with `GetCommand` to work. Seems like the expression mapping only executes until the first `<CR>` to select what to operate on. The remaining normal commands get executed afterwards. Leaving me with the part between the first `-` and the cursor position deleted, and the part between the cursor and the second `-` visually selected.

So here's a full working example of the multiline inner minus text object:
```vim
onoremap <expr> i- GetCommand()
func GetCommand()
    " mode is always 'no' when invoked from omap
    echom 'mode ' . mode(1)
    let [g:lnum1, g:col1] = searchpos('-', 'Wb')
    if g:lnum1 == 0
        return "\<Esc>"
    endif
    let [g:lnum2, g:col2] = searchpos('-', 'W')
    if g:lnum2 == 0
        return "\<Esc>"
    endif
    return ":call Select()\<CR>"
endfunc

func Select()
    call cursor([g:lnum1, g:col1])
    normal! 1 v
    call cursor([g:lnum2, g:col2])
    execute "normal! \<BS>"
endfunc
```
I'm testing it on this file (cursor on `x`):
```
aaa - bbb
x
ddd - eee
```

As indicated by the comment, `mode()` doesn't actually help here to recognize the forced motion. This only works for visual mode (`vmap` and friends), where it can be used to recognize character, linewise or blockwise visual mode. If invoked via omaps, it's always `no`, regardless of whether I used `di-`, `dvi-` or `dVi-` (this last case is the one which currently doesn't work as expected).

----

And there's also other cases where it doesn't work. If you're using the matchit plugin (enabled by default), consider this example:
```go
type Client struct {
	cache cache.Cache
}
```
Put your cursor on `Client` and type `dV%`. I'd expect it to delete the three full lines, but it doesn't. The `V` is ignored again. I had the matchit plugin disabled for a while because of this (without the plugin in works).

----

But there's even a builtin which ignores force motion. Search for a string which appears a few times in your document, then type `dVgn`. I'd expect it to delete the full line which contains the next match, but it only deletes the next match (character wise). If this would work it would be a nice interactive alternative to `:g//d`.

**Comment by brammool** [comment]
> @brammool: I can't get your example with `GetCommand` to work. Seems
> like the expression mapping only executes until the first `<CR>` to
> select what to operate on. The remaining normal commands get executed
> afterwards. Leaving me with the part between the first `-` and the
> cursor position deleted, and the part between the cursor and the
> second `-` visually selected.
> 
> So here's a full working example of the multiline inner minus text object:
> ```vim
> onoremap <expr> i- GetCommand()
> func GetCommand()
>     " mode is always 'no' when invoked from omap
>     echom 'mode ' . mode(1)
>     let [g:lnum1, g:col1] = searchpos('-', 'Wb')
>     if g:lnum1 == 0
>         return "\<Esc>"
>     endif
>     let [g:lnum2, g:col2] = searchpos('-', 'W')
>     if g:lnum2 == 0
>         return "\<Esc>"
>     endif
>     return ":call Select()\<CR>"
> endfunc
> func Select()
>     call cursor([g:lnum1, g:col1])
>     normal! 1 v
>     call cursor([g:lnum2, g:col2])
>     execute "normal! \<BS>"
> endfunc
> ```
> I'm testing it on this file (cursor on `x`):
> ```
> aaa - bbb
> x
> ddd - eee
> ```
> 
> As indicated by the comment, `mode()` doesn't actually help here to
> recognize the forced motion. This only works for visual mode (`vmap`
> and friends), where it can be used to recognize character, linewise or
> blockwise visual mode. If invoked via omaps, it's always `no`,
> regardless of whether I used `di-`, `dvi-` or `dVi-` (this last case
> is the one which currently doesn't work as expected).
> 
> ----
> 
> And there's also other cases where it doesn't work. If you're using
> the matchit plugin (enabled by default), consider this example:
> ```go
> type Client struct {
> 	cache cache.Cache
> }
> ```
> Put your cursor on `Client` and type `dV%`. I'd expect it to delete
> the three full lines, but it doesn't. The `V` is ignored again. I had
> the matchit plugin disabled for a while because of this (without the
> plugin in works).
> 
> ----
> 
> But there's even a builtin which ignores force motion. Search for a
> string which appears a few times in your document, then type `dVgn`.
> I'd expect it to delete the full line which contains the next match,
> but it only deletes the next match (character wise). If this would
> work it would be a nice interactive alternative to `:g//d`.

OK, the "v" or "V" typed in operator pending mode actually doesn't
change the mode.  It's stored in motioin_force, and I don't see an
existing way to get its value.

We could modify mode() to add this info.  Currently we have "no" for
operator-pending, we could att eh motion_force after it.  I think that's
the most consistent way.

-- 
An actual excerpt from a classified section of a city newspaper:
"Illiterate?  Write today for free help!"

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by wellle** [comment]
That sounds good to me. Thank you for moving this forward! 👍

**Comment by chrisbra** [comment]
Can we sum up the conclusion? I'll have a look then.
 - use `mode(1)` and in operator pending mode return in addition `motion_force`?
    This only applies to operator pending mode?
 - @welle If I use your [example](https://github.com/vim/vim/issues/3490#issuecomment-425642966), what do you expect for each of `di-`, `dvi-` and `dVi-` to behave?

**Comment by wellle** [comment]
@chrisbra: Thank you! I'd also suggest to include `force_motion` if it was forced to visual blockwise:

1. `di-`: `mode(1) == 'no'` (not changing behaviour in this case)
2. `dvi-`: `mode(1) == 'nov'`
3. `dVi-`: `mode(1) == 'noV'`
4. `d^Vi-`: `mode(1) == 'no^V'` (where `^V` means a literal `<Ctrl-V>` in both places)

This is assuming that `mode(1)` returns a literal `^V` when called from visual block mode. `:help mode()` is displaying it as `CTRL-V`, but I wasn't able to observe it in some quick testing (how do I call `mode()` without leaving visual mode?).

**Comment by chrisbra** [comment]
> @chrisbra: Thank you! I'd also suggest to include force_motion if it was forced to visual blockwise:

you mean `motion_force`? But to be clear, this should only be output if in operator_pending mode, right?

> 1. ...

Yeah, that's what I expected from our discussion. But I actually meant how the output of the text should look like :)


> This is assuming that mode(1) returns a literal ^V when called from visual block mode. :help mode() is displaying it as CTRL-V, but I wasn't able to observe it in some quick testing (how do I call mode() without leaving visual mode?).

I usually stick in the `mode()` expression into a statusline option, so that one does not change modes to call `:mode()`.

**Comment by wellle** [comment]
> you mean `motion_force`?

Ah yes, of course!

> But to be clear, this should only be output if in operator_pending mode, right?

Yes, in the examples below I meant what the expression mapping in my example would see, which is in operator pending mode.

----

> Yeah, that's what I expected from our discussion. But I actually meant how the output of the text should look like :)

Ah, I see. In the example
```
111
aaa - bbb
x
ddd - eee
222
``` 
with the cursor on `x`, if I type `di-` it just deletes characterwise between the dashes:
```
111
aaa -- eee
222
```
The same should happen if I force it to be character wise (which it already is) with `dvi-`. However, if I type `dVi-` it should force linewise and delete the three full lines:
```
111
222
```
And if I do `d^Vi-` it should actually select the block around the dashes (in this very specific example) and I'd expect to end up with
```
111
aaabbb
x
dddeee
222
```
For reference the same already happens if I search for space, put the cursor after the first dash and type `d^Vn`, which is the same motion in this case and leads to the expected blockwise deletion.

----

> I usually stick in the `mode()` expression into a statusline option, so that one does not change modes to call `:mode()`.

Ah that's very smart! I managed to get `^V` (literal, single character) for visual block mode in this case. (I did not manage to make it show `no` in operator pending mode though, having `set statusline=%!mode(1)` in my vimrc I typed `diw` slowly but only saw `n` all the time.)

**Comment by chrisbra** [comment]
Can you please check this patch: https://github.com/chrisbra/vim/commit/92d8678680419eb789c27c912165b399df25bcde

> (I did not manage to make it show no in operator pending mode though, having set statusline=%!mode(1) in my vimrc I typed diw slowly but only saw n all the time.)

Yeah I noticed that. Possibly missing a screen redraw here (that would also update the statusline). But that is another topic  😳

**Comment by wellle** [comment]
@chrisbra: Thanks a lot, I just tested it with this updated code:

```vim
onoremap <expr> i- GetCommand()

func GetCommand()
    let g:motion_force = mode(1)[2]

    let [g:lnum1, g:col1] = searchpos('-', 'Wb')
    if g:lnum1 == 0
        return "\<Esc>"
    endif
    let [g:lnum2, g:col2] = searchpos('-', 'W')
    if g:lnum2 == 0
        return "\<Esc>"
    endif
    return ":call Select()\<CR>"
endfunc

func Select()
    " select characterwise
    call cursor([g:lnum1, g:col1])
    normal! 1 v
    call cursor([g:lnum2, g:col2])
    execute "normal! \<BS>"

    " potentially force linewise/blockwise
    if g:motion_force ==# "v"
        echom 'ignoring characterwise'
    elseif g:motion_force ==# "V"
        echom 'forcing linewise'
        normal! V
    elseif g:motion_force ==# "^V"
        echom 'forcing blockwise'
        execute "normal! \<C-V>"
    endif
endfunc
```

Your change works as expected and with the code above the text object now works as expected when forced linewise. 🎉 

Interestingly forcing blockwise doesn't seem to work though. I experimented a bit and my feeling is that this is unrelated to your change though. If I put `execute "normal! \<C-V>"` into the `"V"` case, then I can force it to behave blockwise by typing `dVi-`.

With the version above (with `<C-V>` in the `<C-V>` case), instead it seems to select blockwise from the character the command was started on (`x` in our case) and the end of the selection. While it should do blockwise from start of selection to end of selection.

And I also tried that (`d^Vi-`) in Vim without your commit and it behaves in the same way. Looks like `o_CTRL-V` is already setting the start of the visual selection in a way so that the omap can't set it itself anymore.

Since it seems unrelated, and my initial request is fulfilled (I mostly care about forcing linewise) I'm fine with keeping it as it is and move forward with your change. I'm happy to open a separate issue about forcing blockwise if you want. If fixing forcing blockwise is easy, I'm of course also happy to have this included here, but I can't judge that.

Thanks again already for making linewise work! 👍

**Comment by andymass** [comment]
@wellle Maybe the issue you saw is due to the faulty check `==# "^V"`?  should be literal ctrl-v.

@chrisbra thanks for working on this!

**Comment by chrisbra** [comment]
There was one problem with testing blockwise mode, as @andymass pointed out. Check for `g:motion_force ==# "\<c-v>"`

Ah, I see the other problem. In `do_pending_operator` we do explicitly set the start position of the visual mode to the start of the operator, if `Ctrl-V` has been pressed (this does not happen for characterwise or linewise visual mode) and this kind of defeats the whole purpose of your mapping. We might be able to only set it, if visual mode is not active. (I hope this does not break anything).

So all together the patch should look like this: 
https://github.com/vim/vim/compare/master...chrisbra:motion_force?expand=1

I actually stumbled upon the problem yesterday when creating the testcase, but thought I made a mistake with the test. :(

**Comment by wellle** [comment]
Ah sorry, I actually had a literal `<C-V>` in my comparison, by copy pasting it here it became `^V` (two characters). The escaped way is more clear though, thanks!

I just tried it again with your updated patch, and it works perfectly! 🎉 

Thank you very much for you work! 👍

**Comment by liushapku** [comment]
Recently, In a plugin I am writing, I also get the needs to allow `motion_force` be effective when an visual area is selected using Ex commands in operator-pending mode; for which I asked in [vim.stackexchange](https://vi.stackexchange.com/questions/18099/overwrite-motion-wiseness-of-operator-pending-mode/18100#18100). I feel the approach provided here to add the `motion_force` to the return value of `mode(1)` is the most natural choice. Hope this patch can be added to vim in the near future.

**Comment by wellle** [comment]
@chrisbra: Any update here?

**Comment by chrisbra** [comment]
there is nothing for me left to do

**Comment by wellle** [comment]
Ok thanks!

@brammool: Do you think we can schedule this for release?
